### PR TITLE
Improve "Failed to save the story" message

### LINKED ISSUE #11082: Improve "Failed to save the story" message
## Feature Description

Splitting this out of #10963

If we fail to save the story, we do not currently provide any context about why that happens.

For example, if the REST API returns a 5xx or 4xx status, we should be able to say that in the message.

Let's:

* Get the `err` from the `catch` and see if we can enhance the snackbar message. Maybe with `err.message`?
* Call `console.log(err)` so it's easier to spot this in dev tools; helps with support
* Call `trackError()` so we get telemetry on these errors.

Here's where we display the message:

https://github.com/GoogleForCreators/web-stories-wp/blob/3cc4ee5c5c02eb9dc2e652660a5d7a731aa3babe/packages/story-editor/src/app/story/actions/useSaveStory.js#L108-L113

We could even look into adding an e2e test for this, with a test plugin that mocks typical scenarios like 5xx or 4xx status. Like we do for status check.

### PR DESCRIPTION
**Pull request by timarney**
## Context

Adds more specific error message for when a story fails to save.

See https://github.com/GoogleForCreators/web-stories-wp/issues/11082#issue-1180689949 for more details

## Summary

Updates failed save story to handle 400, 401, 403, 500 errors + track errors

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No 

### Does this PR change what data or activity we track or use?

Yes --- tracks error here 
https://github.com/GoogleForCreators/web-stories-wp/pull/11144/files#diff-268fab14c97529f4d480ac6c70e2082129e453aceff794f2237dcc6e3bad9180R152

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11082

### CODE REVIEW

**Comment by timarney** [comment]
@spacedmonkey @swissspidy 

I'm having an issue getting an e2e test running properly for this ...

For real testing if I force an error from the `Stories_Controller.php` i.e.

Screen Shot 2022-04-04 at 7 31 07 AM

I get the expected result (SnackBar message with a server error)

| Response  |      Result    |  
|----------|:-------------:|
| real-response |  Screen Shot 2022-04-04 at 7 24 24 AM | 


<hr>

In Jest I can see the Mock response I setup seems to be working but no SnackBar message shows --- I also see Jest has a expected not to 500 error.  Is there something for Jest I need to adjust to allow the 500 in this case?


### Jest 

| Response  |      Result    |  
|----------|:-------------:|
| test-response |  jest|

**Comment by swissspidy** [comment]
Apologies for the delay here.

> Is there something for Jest I need to adjust to allow the 500 in this case?

Yes. See 6caa3ab8e214cdabd036dd3602a5a37474bc9d02

**Explanation:** 

We use a package called [`@wordpress/jest-console`](https://npmjs.com/package/@wordpress/jest-console) to automatically surface browser console messages to Jest. This allows us to catch potentially unexpected errors (which would indicate a bug). In this case of course we are expecting this log message because we are forcing it ourselves.

I also took the liberty of refining the snackbar messages a little bit.

Not sure how well it works because I don't think we really have access to the HTTP status code in the `catch` block right now.

**Comment by timarney** [comment]
Rebased this to latest main

File: packages/e2e-tests/src/specs/dashboard/dashboard.js [hunk 1]
```diff
@@ -43,8 +46,12 @@ describe('Stories Dashboard', () => {
   });
 
   it('should choose sort option for display', async () => {
+    // dropdown needs a story for filtering
```
Comment (timarney) [inline packages/e2e-tests/src/specs/dashboard/dashboard.js]: Noticed the dashboard test was failing ...

The dropdown didn't have a story to work with.

Comment (timarney) [reply packages/e2e-tests/src/specs/dashboard/dashboard.js]: I cherry picked that commit for this PR

https://github.com/GoogleForCreators/web-stories-wp/pull/11227

**Comment by swissspidy** [comment]
While debugging an issue reported via support, I noticed that errors thrown in `getStoryMarkup` (where we assemble the full HTML for a story) were not caught. I just fixed this now in 62d1df4d9b51ba030e7fe28231f2c9549368a8c2 as part of this PR. Works like a charm:

Screenshot 2022-04-11 at 15 37 10

**Review by swissspidy — APPROVED** [review]

**Pull request merged**